\documentclass[12pt,a4paper]{article}
%\usepackage[margin=1in]{geometry}

% Localization and encoding
\usepackage[utf8]{inputenc}
\usepackage[english]{babel}
\usepackage{csquotes}

% Math
\usepackage{amsthm}
\usepackage{amsmath}
\usepackage{amsfonts}

% Graphics
%\usepackage{rotating}
\usepackage{graphicx}
\graphicspath{ {./resources/images/} }
\usepackage{multirow}
\usepackage[linewidth=1pt]{mdframed}

% algorithms
%\usepackage{algorithm}
%\usepackage{algorithmic}
\usepackage[algoruled]{algorithm2e}

\usepackage{pdfpages}
\usepackage{multirow}
\usepackage{placeins}
\usepackage{hyperref}
\usepackage{caption}
\usepackage{tikz}
\usetikzlibrary{math}
\usetikzlibrary{calc}
\usetikzlibrary{arrows.meta}



% Counters
\newcounter{thmCounter}
\newcounter{defCounter}

% Theorem environments
\newtheorem{observation}{\bf Observation}
\newtheorem{theorem}{\bf Theorem}
\newtheorem{lemma}{\bf Lemma}
\newtheorem{prop}{\bf Proposition}
\newtheorem{rmk}{\bf Remark}
\newtheorem{definition}{\bf Definition}
\newtheorem{example}{\bf Example}
\newtheorem{corollary}{\bf Corollary}

\newcommand{\TODO}[1]{\textcolor{red}{#1}}



\sloppy
\title{Haplotype Inference via Boolean Satisfiability and Integer Linear Programming}
\date{2020 Autumn}
\author{Jarkko Savela}

\begin{document}
\pagenumbering{gobble}
\maketitle
\newpage
\tableofcontents
\newpage
\pagenumbering{arabic}

\section{Introduction}
Haplotype inference is a crucial problem of modern genetics with wide-ranging applications in, e.g., correlating disease susceptibility to individuals' haplotypes.
Due to technical difficulties the direct measurement of haplotypes is challenging, and current methods rely on indirectly inferring haplotypes from genotype data.

In general there are many different haplotypes that can result in the same genotype, 
which is why a criterion is needed to choose between the possibilities when inferring haplotype from genotype.
Several criteria have been suggested such as perfect phylogeny as well as pure parsimony~\cite{gusfield2019integer}.
The pure parsimony criterion has been experimentally validated to yield relatively accurate haplotypes.

Different methods of solving the haplotype inference problem have been suggested including customized branch-and-bound algorithms, integer linear programs (ILPs) as well as Boolean satisfiability (SAT) encodings and SAT-based methods utilizing pseudo-boolean optimization and answer-set programming.

This report focuses on SAT and ILP formulations to the problem of haplotype inference by pure parsimony (HIPP) as presented by Lynce and Marques-Silva~\cite{DBLP:conf/aaai/LynceM06}, Gusfield~\cite{DBLP:conf/cpm/Gusfield03} as well as Brown and Harrower~\cite{DBLP:journals/tcbb/BrownH06}.
We begin with a brief overview of Boolean satisfiability and integer linear programming in Sections~\ref{sec:sat} and~\ref{sec:ilp}, respectively.
In Section~\ref{sec:hipp} we present the HIPP problem formally.
Section~\ref{sec:hipp-sat} presents the SAT encoding whereas 
Section~\ref{sec:ilp-hipp} presents the two different ILP formulations. 



\section{Boolean Satisfiability}
\label{sec:sat}
Boolean satisfiability~\cite{DBLP:series/faia/2009-185}, often referred to as the SAT problem, is the satisfiability problem of propositional logic.
That is, the problem is to find an assignment $s$ that satisfies the input formula $\phi$.
The formula $\phi$ is inductively constructed from propositional atoms (i.e., Boolean variables) and the usual connectives; conjunction $\wedge$, disjunction $\vee$, negation $\neg$, implication $\rightarrow$ and bi-implication $\leftrightarrow$.
The assignment $s$ gives each boolean variable in $\phi$ a truth value, 0 or 1.

In practice the input formula is usually restricted to be in \emph{conjunctive normal form} (CNF).
A formula $\phi$ is in CNF if it is a conjunction of \emph{clauses}
$$ \bigwedge_{i\in I} C_i $$
where each clause $C_i$ is a disjunction of \emph{literals}
$$ C_i = \bigvee_{j\in J_i} l_j $$
and each literal is either an atom $p$ or a negated atom $\neg p$.
CNF is a convenient standard format partly due to the fact that efficient polynomial-time algorithms exist for transforming arbitrary formulas into \emph{linear-size, equisatisfiable} CNF formulas~\cite{tseitin1983, DBLP:journals/jsc/PlaistedG86}.

Practically efficient algorithms for solving the SAT problem have been developed despite the fact that SAT is an NP-Complete problem~\cite{DBLP:conf/stoc/Cook71}, 
and SAT has become an important problem solving technique with applications in model checking and verification, planning, scheduling, cryptography, computational biology et cetera.
The most efficient current SAT solvers are based on the \emph{conflict-driven clause learning} algorithm (CDCL) and use many optimizations and efficient data structures~\cite{DBLP:conf/iccad/SilvaS96, DBLP:journals/tc/Marques-SilvaS99, DBLP:conf/dac/MoskewiczMZZM01, DBLP:conf/aaai/GomesSK98, DBLP:journals/dam/GoldbergN07, ryan2004efficient, DBLP:conf/aaai/BayardoS97, DBLP:conf/sat/LewisSB05}.


\section{Integer Linear Programming}
\label{sec:ilp}
Integer linear programming (ILP) is the problem of finding an optimal assignment to a set of integer variables subject to constraints in the form of linear inequalities~\cite{DBLP:books/ph/PapadimitriouS82}.
An ILP program thus consists of a vector of integer variables $\mathbf{x}$, a cost vector $\mathbf{c}$ as well as a matrix $A$ and vector $\mathbf{b}$ of coefficients with the goal of minimizing $\mathbf{c}^T \mathbf{x}$ under the constraint that $A\mathbf{x} \leq \mathbf{b}$.

ILP is also an NP-hard problem~\cite{DBLP:conf/coco/Karp72} and finds applications in planning, scheduling, computational biology and in many other domains.

The \emph{simplex algorithm} is a practically efficient (although worst-case exponential time) algorithm for solving general linear programs where the variables may take non-integral values as well.
ILPs can be solved via a \emph{branch-and-bound} procedure in which the simplex algorithm is used to solve the continuous relaxations of the program with or without the use of \emph{cutting planes}.
Branch-and-bound procedure utilizing cutting planes, i.e., \emph{branch-and-cut}, uses auxiliary constraints to prune away non-integral solutions~\cite{DBLP:books/ph/PapadimitriouS82}.


\section{Haplotype inference}
\label{sec:hipp}
\TODO{How relevant is the 2-allele haplotype inference? How to generalize to several alleles?}

The human genetic information is contained in the double-stranded DNA molecules consisting of a linear sequence of 4 bases (\textbf{A}denosine, \textbf{C}ytosine, \textbf{G}uanine, \textbf{T}hymine) in the phosphate-deoxyribose backbone.
The DNA molecules, which are packed into \emph{chromosomes}, consist of protein-coding intervals, which are called genes, and other non-coding parts.
Different variants of a gene are called \emph{alleles}.
Humans are a \emph{diploid} species, which means that we have two copies of each chromosome (one from each parent).
We thus have two alleles of each gene, and we are said to be \emph{homozygous} with regard to a gene if we have two identical alleles and \emph{heterozygous} otherwise.

\emph{Haplotype} refers to the information contained in a single chromosome whereas the conflated information contained in the pair of  homologous chromosomes inherited is referred to as the \emph{genotype}.
In practice it is difficult and expensive to experimentally detect haplotype information.
Detecting genotype information has, however, become much faster, cheaper and more reliable in the recent decades.
Potential haplotypes can be computationally inferred from genotype data, and this is the problem referred to as the haplotype inference (HI) problem.
Haplotype data has the potential to yield new insight into genetic predisposition to, e.g., diseases, which makes the HI problem of great importance.

\subsection{Haplotype Inference by Pure Parsimony}
The genotype data for the computational problem is represented as vectors over the alphabet $\{0, 1, 2\}$. 
The values $0$ and $1$ indicate that the location in the diploid genotype is homozygous with allele $0$ or allele $1$, respectively, 
whereas the value $2$ indicates that the location is heterozygous.
The corresponding haplotype data then consists of a pair of vectors over the binary alphabet $\{0, 1\}$ representing the maternal and paternal haplotypes.
Denote a genotype vector by $g=g_1\dots g_m$ and the haplotype vectors $h=h_1\dots h_m$ and $h'=h'_1\dots h'_m$.
The pair of haplotype vectors $(h, h')$ corresponds to genotype $g$ if 
\begin{itemize}
\item $h_i=h'_i=0$ whenever $g_i = 0$,
\item $h_i=h'_i=1$ whenever $g_i = 1$, and
\item $h_i \neq h'_i$ whenever $g_i = 2$.
\end{itemize}
The haplotype pair $(01011, 00001)$, for example, corresponds to the genotype $02021$.
Note that there can be several distinct pairs of haplotype vectors for each genotype.
For example the genotype $02021$ corresponds also to the pair $(00011, 01001)$.
In fact the number of different haplotype pairs for $g$ is $2^{k-1}$, where $k$ is the number of heterozygous loci, i.e., the number of 2s in the string.

The input to the HI problem is the genotype data of $n$ individuals with the genotype vectors being all of length $m$.
The objective is to find a pair of haplotype vectors for each of the $n$ genotypes such that the two haplotype vectors yield exactly the genotype vector.
That is, for each genotype vector $g$ find a pair of haplotype vectors $(h, h')$ such that $g$ corresponds to $(h, h')$.
Without additional constraints the problem is easy to solve but yields no guarantee that the chosen haplotypes are likely to be the correct ones.
The Haplotype Inference by Pure Parsimony (HIPP) sets another constraint on this problem, namely that the set of distinct haplotype vectors used should be as small as possible.

\TODO{How does $n$ affect the accuracy?}

\TODO{Elaborate: "relatively accurate haplotype data"}

This condition has been experimentally validated to yield relatively accurate haplotype data~\cite{DBLP:journals/bioinformatics/WangX03a}.
The HIPP problem was shown by Lancia et al to be APX-hard~\cite{DBLP:journals/informs/LanciaPR04}.
\TODO{Need to mention what the complexity class APX is!}

In the following we present three solutions for the HIPP problem: a SAT encoding by Lynce and Marques-Silva~\cite{DBLP:conf/aaai/LynceM06} and ILP formulations by Gusfield~\cite{DBLP:conf/cpm/Gusfield03} and by Brown and Harrower~\cite{DBLP:journals/tcbb/BrownH06}.

\section{SAT formulation}
\label{sec:hipp-sat}
\TODO{Clarify what each subscripted variable means to make reading easier.}

Denote by $\mathcal{G}=\{g_i\mid 1\leq i \leq n\}$ a set of $n$ genotype vectors of length $m$ and by $g_{i,j}$ the $j$'th entry of genotype vector $g_i$.
We construct a propositional formula $\phi (\mathcal{G}, r)$ which is satisfiable if and only if there exists a set of haplotypes $\mathcal{H}$ of size $r$ such that each genotype $g_i\in \mathcal{G}$ is explained by some pair of haplotypes $(h_j, h_k)$ in $\mathcal{H}$.
To solve the HIPP problem the satisfiability of $\phi (\mathcal{G}, r)$ is checked for increasing values of $r$ until a value is found such that $\phi (\mathcal{G}, r)$ is satisfiable but $\phi (\mathcal{G}, r-1)$ is not.
This value of $r$ is then the minimum number of haplotypes required, and the actual haplotypes can be extracted from the satisfying assignment to $\phi (\mathcal{G}, r)$.

The formula $\phi (\mathcal{G}, r)$ considers a set of $r$ distinct haplotypes $h_k$ and encodes the assignment of a pair $(h_{k_1}, h_{k_2})$ to each $g_i$ using Boolean variables $s^a_{k_1,i}$, $s^b_{k_2,i}$ and $h_{k,j}$.
That is, variables $s^a_{k_1,i}$ and $s^b_{k_2,i}$ being set true indicates that $(h_{k_1}, h_{k_2})$ is the pair of haplotypes assigned to genotype $g_i$.
The variable $h_{k,j}$ is the $j$'th entry of haplotype $h_k$.
The formula then requires two types of constraints
\begin{enumerate}
\item[(i)] The selector variable $s^a_{k,i}$ should select a unique haplotype, i.e., for each genotype $g_i$ exactly one variable $s^a_{k,i}$ should be set true (and similarly for $s^b_{k,i}$). 
\item[(ii)] Each genotype $g_i$ should be explained by its selected pair of haplotypes $(h_{k_1}, h_{k_2})$. 
\end{enumerate}

Constraint (i) corresponds to equations $\sum_{1\leq k\leq r} s^a_{k,i} = 1$ and $\sum_{1\leq k\leq r} s^b_{k,i} = 1$ holding for all $i$ such that $1\leq i \leq n$.
This constraint is enforced in $\phi (\mathcal{R}, r)$ utilizing a propositional encoding of an \emph{1-bit adder circuit}.
The adder circuit uses variables $v^a_{k,i}$ where $1\leq k \leq r+1$ and $1\leq i\leq n$ and the following constraints (the ones for $s^b_{k,i}$ are similar).
\begin{eqnarray}
\forall i\text{ such that } 1\leq i\leq n: & v^a_{1,i} \leftrightarrow s^a_{1,i}\label{hipp-sat:clause1}\\
\forall k, i\text{ such that } 2\leq k \leq r,\ 1\leq i\leq n: & \neg v^a_{k,i} \vee \neg s^a_{k,i}\label{hipp-sat:clause2}\\
\forall k, i\text{ such that } 1\leq k \leq r,\ 1\leq i\leq n: & v^a_{k+1,i} \leftrightarrow (s^a_{k,i}\vee v^a_{k,i} )\label{hipp-sat:clause3}\\
\forall i\text{ such that } 1\leq i\leq n: & v^a_{r+1,i}=1\label{hipp-sat:clause4}
\end{eqnarray}
Constraint \eqref{hipp-sat:clause1} enforces $v^a_{1,i}$ and $s^a_{1,i}$ to take on the same value in every satisfying assignment whereas constraint \eqref{hipp-sat:clause4} enforces variable $v^a_{r+1,i}$ to be evaluated to true.
Constraint \eqref{hipp-sat:clause2} prevents $v^a_{k,i}$ and $s^a_{k,i}$ from being simultaneously assigned true, 
and constraint \eqref{hipp-sat:clause3} sets $v^a_{k+1,i}$ to 1 whenever either $s^a_{k,i}$ or $v^a_{k,i}$ is assigned 1.

It can be shown using induction that the variable $v^a_{k,i}$ (for $k>1$) represents the disjunction $\bigvee_{1\leq h\leq k-1}s^a_{h,i} = s^a_{1,i}\vee\dots \vee s^a_{k-1,i}$.
From constraints \eqref{hipp-sat:clause1} and \eqref{hipp-sat:clause3} it follows that $v^a_{2,i}$ is equivalent to $s^a_{1,i}$.
Constraint \eqref{hipp-sat:clause3} also provides the proof of the inductive step.
Thus, constraint \eqref{hipp-sat:clause4} which sets $v^a_{r+1,i}=1$ ensures that at least one of $s^a_{1,i},\dots , s^a_{r,i}$ is set to 1.
Furthermore, there can only be one $s^a_{k,i}$ set to 1 due to the following reasoning.
Since $v^a_{k,i}$ is equivalent to $\bigvee_{1\leq h\leq k-1}s^a_{h,i} = s^a_{1,i}\vee\dots \vee s^a_{k-1,i}$ it follows that $v^a_{k,i}$ being set 1 enforces all $v^a_{h,i}$ to be set 1 for $h>k$.
Now assume there are two indices $h$ and $h'$ ($h<h'$) such that both $s^a_{h,i}$ and $s^a_{h',i}$ are assigned true.
By the previous observation $v^a_{h',i}$ is also assigned true which violates constraint \eqref{hipp-sat:clause2}.

Next we formalize constraint (ii) stating that each genotype $g_i$ should be explained by its selected pair of haplotypes $(h_{k_1}, h_{k_2})$.
If $g_{i,j}$ (the $j$'th entry of $g_i$) is 0, then the $j$'th entries of the chosen haplotypes should also be 0.
This is encoded by $s^a_{k,i}\rightarrow\neg h_{k,j}$ and $s^b_{k,i}\rightarrow\neg h_{k,j}$ for all $1\leq k \leq r$.
Similarly, if $g_{i,j}$ is 1, then the $j$'th entries of the chosen haplotypes should also be 1, which is encoded by $s^a_{k,i}\rightarrow h_{k,j}$ and $s^b_{k,i}\rightarrow h_{k,j}$ for all $1\leq k \leq r$.
If $g_{i,j}$ is 2, then the $j$'th entry of the haplotypes should differ.
Variables $g^a_{i,j}$ and $g^b_{i,j}$ are used to decide which haplotype should be assigned 0 and which should be assigned 1.
The clauses $(g^a_{i,j}\vee g^b_{i,j})\wedge(\neg g^a_{i,j}\vee \neg g^b_{i,j})$ enforce $g^a_{i,j}$ and $g^b_{i,j}$ to take on different values.
The constraints $s^a_{k,i}\rightarrow (h_{k,j}\leftrightarrow g^a_{i,j})$ and $s^b_{k,i}\rightarrow (h_{k,j}\leftrightarrow g^b_{i,j})$ then enforce the selected haplotypes to take on values dictated by $g^a_{i,j}$ and $g^b_{i,j}$.
\begin{figure}
\centering
\small
\begin{mdframed}
\begin{align}
\text{for all  }& i\in\{1,\dots , n\}:\nonumber\\
& \quad\quad\quad\quad \sum_{1\leq k \leq r} s^a_{k,i} = 1 \\
& \quad\quad\quad\quad \sum_{1\leq k \leq r} s^b_{k,i} = 1 \\
\text{for all  }& i\in\{1,\dots , n\},\ k\in\{1,\dots r\}\text{ and }\nonumber\\&\quad\quad j\in\{1,\dots ,m\} \text{ such that }g_{i,j}=0:\nonumber\\
& \quad\quad\quad\quad s^a_{k,i}\rightarrow \neg h_{k,j} \\
& \quad\quad\quad\quad s^b_{k,i}\rightarrow \neg h_{k,j} \\
\text{for all  }& i\in\{1,\dots , n\},\ k\in\{1,\dots r\}\text{ and }\nonumber\\&\quad\quad j\in\{1,\dots ,m\} \text{ such that }g_{i,j}=1:\nonumber\\
& \quad\quad\quad\quad s^a_{k,i}\rightarrow h_{k,j} \\
& \quad\quad\quad\quad s^b_{k,i}\rightarrow h_{k,j} \\
\text{for all  }& i\in\{1,\dots , n\},\ k\in\{1,\dots r\}\text{ and }\nonumber\\&\quad\quad j\in\{1,\dots ,m\} \text{ such that }g_{i,j}=2:\nonumber\\
& \quad\quad\quad\quad g^a_{i,j}\vee g^b_{i,j} \\
& \quad\quad\quad\quad \neg g^a_{i,j}\vee \neg g^b_{i,j} \\
& \quad\quad\quad\quad s^a_{k,i}\rightarrow (h_{k,j}\leftrightarrow g^a_{i,j}) \\
& \quad\quad\quad\quad s^b_{k,i}\rightarrow (h_{k,j}\leftrightarrow g^b_{i,j}) \\
\end{align}
\end{mdframed}
\caption{The plain HIPP-SAT encoding.}
\label{fig:enc-hipp-sat}
\end{figure}

The encoding is presented in Figure~\ref{fig:enc-hipp-sat}.


Lynce and Marques-Silva use two techniques to facilitate and speed up the use of their HIPP-SAT encoding~\cite{DBLP:conf/aaai/LynceM06}.
There are symmetries present in the plain HIPP-SAT encoding, but fortunately these symmetries can be dispatched by utilizing symmetry-breaking predicates.
Since the usage of the encoding consists of calling $\phi (\mathcal{G}, r)$ with values of $r$ from some lower bound $lb$ up to $2n$ until the formula becomes satisfiable, the overall effort can be reduced by finding a large $lb$.
The computation of lower bounds of $r$ are thus useful in speeding up SAT-based haplotype inference.

Let $h$ and $h'$ be two concrete haplotype vectors.
Assume that the valuation of $h_{k_1}$ is $h$, the valuation of $h_{k_2}$ is $h'$ and that $s_{k_1, i}^a$ and  $s_{k_2, i}^b$ are set true.
This means that genotype $g_i$ is explained by haplotypes $h$ and $h'$.
Another equivalent solution to the HIPP-SAT encoding can be constructed from this one by swapping the roles of $h_{k_1}$ and $h_{k_2}$ as well as those of $s_{k_1, i}^a$ and  $s_{k_2, i}^b$.
This is an example of a symmetry present in the plain formulation, 
which can be broken by enforcing the used haplotype to be in lexicographic order.
The ordering constraint should thus enforce that for any valuation $v$ it holds that $v(h_1) < v(h_2) < \dots < v(h_r)$ where $v(h_i)$ is the binary string $v(h_{i,1})v(h_{i,2})\dots v(h_{i,n})$.
Frisch et al~\cite{DBLP:conf/cp/FrischHKMW02} present ways of encoding these types of ordering constraints.
Essentially constraints are added to the encoding that for each $i\in\{1,\dots ,r-1\}$ enforce that $v(h_i)$ is lexicographically less than $v(h_{i+1})$.

To find a lower bound $lb$ Lynce and Marques-Silva construct an \emph{incompatibility graph} of the genotypes~\cite{DBLP:conf/aaai/LynceM06}.
Two genotypes are \emph{incompatible} if there is a locus at which they are both homozygous but with different alleles.
The incompatibility graph has a node for each genotype and an edge between two nodes if the corresponding genotypes are incompatible.
Assume that a clique of size $k$ exists in the constructed incompatibility graph.
Then at least $2k - \sigma$ haplotypes are required where $\sigma$ is the number of genotypes in the clique without heterozygous loci.
The number $2k-\sigma$ can thus be taken as the lower bound $lb$ for $r$.
Since finding the largest clique in a graph is NP-hard it can be more beneficial to find a \emph{maximal} clique using greedy heuristics.


\section{ILP formulations}
\label{sec:ilp-hipp}
Two different ILP formulations of the HIPP problem, RTIP~\cite{DBLP:conf/cpm/Gusfield03} and PolyIP~\cite{DBLP:journals/tcbb/BrownH06}, are presented in this section.
RTIP is an exponential-size encoding formulated and empirically evaluated by Gusfield in~\cite{DBLP:conf/cpm/Gusfield03}.
Similar polynomial-size encodings were simultaneously and independently formulated~\cite{DBLP:journals/tcbb/BrownH06, DBLP:conf/recomb/HalldorssonBELYI02, DBLP:journals/informs/LanciaPR04}.
Out of the polynomial size encodings we showcase PolyIP which is formulated by Brown and Harrower~\cite{DBLP:journals/tcbb/BrownH06}.

\subsection{RTIP}

The RTIP encoding contains variables representing the individual haplotype vectors as well as ones representing certain pairs of haplotype vectors (which correspond to a certain genotype).
Denote by $k_i$ the numbers of heterozygous loci (i.e., number of 2s in the vector) in genotype $g_i$.
Now there are $2^{k_i-1}$ distinct haplotype pairs corresponding to genotype $g_i$.
Denote by $L$ the number of distinct haplotypes appearing in the haplotype pairs of all $g_i\in\mathcal{G}$.
The encoding contains variables $y_{i,j}$ where $1\leq j \leq 2^{k_i-1}$ representing each of the possible haplotype pairs corresponding to $g_i$.
Similarly, each of the individual haplotype vectors appearing in the pairs is assigned a variable $x_l$.
Both the $y_{i,j}$ and $x_l$ variables are restricted to the range $[0,1]$, i.e. they may take only two distinct values, 0 or 1 (since only integer solutions are sought).
Setting $y_{i,j}$ to 1 indicates that $g_i$ is resolved using haplotype pair $p_j$ whereas 
setting $x_l$ to 1 indicates that haplotype $h_l$ is being used in some pair to resolve a genotype.
The constraint
$$ \text{for all  }i\in\{1,\dots , n\}:\quad \sum_{1\leq j \leq 2^{k_i-1}} y_{i,j} = 1 $$
ensures that each genotype is resolved by exactly one haplotype pair $p_j$.
Denote the haplotypes in pair $p_j$ by $h_{l(j,a)}$ and $h_{l(j,b)}$.
To ensure that the haplotypes of a used pair are marked used the encoding contains the following inequalities for all $i$ ($1\leq i \leq n$) and $j$ ($1\leq j \leq 2^{k_i-1}$).
\begin{eqnarray*}
y_{i,j} - x_{l(j,a)} & \leq & 0 \\
y_{i,j} - x_{l(j,b)} & \leq & 0 \\
\end{eqnarray*}
The objective of RTIP is to minize the number of used haplotypes, i.e., $\sum_{1\leq l \leq L} x_l$.
The entire encoding is presented in Figure~\ref{fig:enc-rtip}.

\begin{figure}
\begin{mdframed}
\centering
Minimize $\sum_{1\leq l \leq L} x_l$
\begin{align}
\text{for all  }i\in\{1,\dots , n\}:&\quad \sum_{1\leq j \leq 2^{k_i-1}} y_{i,j} = 1 \\
\text{for all  }i\in\{1,\dots , n\}, j\in\{1,\dots , 2^{k_i-1}\}:& \quad y_{i,j} - x_{l(j,a)}  \leq  0 \\
\text{for all  }i\in\{1,\dots , n\}, j\in\{1,\dots , 2^{k_i-1}\}:&\quad y_{i,j} - x_{l(j,b)}  \leq  0 \\
\text{for all  }l\in\{1,\dots , L\}:&\quad x_l \in \{0,1\} \\
\text{for all  }i\in\{1,\dots , n\}, j\in\{1,\dots , 2^{k_i-1}\}:&\quad y_{i,j} \in \{0,1\} 
\end{align}
\end{mdframed}
\caption{The RTIP encoding.}
\label{fig:enc-rtip}
\end{figure}

The RTIP encoding is of exponential size in the worst case since  the number of $y_{i,j}$ variables is exponential with respect to the number of $2$s in the genotype vectors.
This restricts the usability of this encoding to very small problem instances.
Gusfield, however, shows that two optimizations significantly reduce the size of the encoding making it usable for instance sizes of practical interest.

The first optimization reduces the number of $y_{i,j}$ and $x_l$ variables used in the encoding.
Let $p_j=(h_{l(j,a)}, h_{l(j,b)})$ be a haplotype pair corresponding to genotype $g_i$.
If neither $h_{l(j,a)}$ nor $h_{l(j,b)}$ appears in any other haplotype pair (for any genotype of the input), 
then pair $p_j$ is excluded from the encoding, i.e., 
variables $y_{i,j}$, $x_{l(j,a)}$ and $x_{l(j,b)}$ are not included in the encoding.
In the case that no haplotype pair remains for $g_i$, an optimal solution to the original problem can be constructed from an optimal solution to the reduced problem by choosing an arbitrary pair for $g_i$ which is bound to increase cost by 2.
In the case that $g_i$ does have haplotype pairs remaining, an optimal solution to the reduced problem will correspond to an optimal solution of the original problem with the excluded $y$ and $x$ variables set to $0$.

The second optimization concerns the construction of the encoding.
Enumerating all of the $y$ and $x$ variables and removing the unnecessary ones in line with the first optimization still incurs the cost of generating all potential $y$ and $x$ variables which can become a bottleneck.
This can be mitigated by generating only the haplotypes from intersections as follows.
We denote the set of haplotypes in the potential haplotype pairs of $g_i$ by $H_i$.
The idea is to let $(i,i')$ range over the possible pairs of genotype indices and generate variables for haplotypes in intersections $H_i\cap H_{i'}$.
This suffices to generate all the variables for haplotypes and pairs of haplotypes in the reduced form of RTIP since exactly the haplotypes appearing in at least two haplotype pairs are included.

Let $g_1$ and $g_2$ be two genotypes whose sets of potential haplotypes are $H_1$ and $H_2$.
To generate the variables for haplotypes ($x_l$) and haplotype pairs ($y_{i,j}$) in the intersection $H_1\cap H_2$ we iterate over the length of the genotype vectors $g_1$ and $g_2$.
If an index is encountered at which $g_1$ has a 0 and $g_2$ a 1 (or vice versa) the intersection $H_1\cap H_2$ must be empty.
Whenever the other entry is 2 and the other one 0 (or 1) we know that all haplotypes in the intersection must contain a 0 (or 1) in this location.
If, on the other hand, an index is found where both $g_1$ and $g_2$ contain a 2, the intersection must contain haplotypes differing at this index.
The indices at which both $g_1$ and $g_2$ have entry 2 are recorded and the haplotypes with every possible combination of $0/1$ at these indices are generated along with the $x$ variables for these haplotypes.
Since each one of these haplotypes has a unique pair the $y$ variables can be simultaneously generated.

\subsection{PolyIP}
In this subsection we present the PolyIP encoding for HIPP as detailed by Brown and Harrower~\cite{DBLP:journals/tcbb/BrownH06}.
The exponential size of the RTIP encoding results from the fact that the haplotypes and pairs thereof are explicitly generated and the ILP encoding has no access to the internal structure of these haplotypes.
That is, the representation of haplotypes in RTIP is atomic.
A polynomial size encoding can be formulated for the HIPP problem by using a structured representation for the haplotypes, i.e., allowing the ILP encoding access to the internal structure of the haplotypes.

Again let $\mathcal{G}=\{g_i\mid 1\leq i \leq n\}$ be a set of $n$ genotype vectors of length $m$ and denote by $g_{i,j}$ the $j$'th entry of genotype vector $g_i$.
Brown and Harrower use a slightly different notation for the genotype vectors to facilitate the formulation of the ILP encoding.
Namely, the interpretation of the entries of the genotype vectors are different.
In Brown and Harrower's notation an entry of 0 denotes a homozygous location with allele 0 and 2 denotes a homozygous location with allele 1 whereas 
an entry of 1 in the genotype vector denotes a heterozygous location.
The benefit of this notation is that now the elementwise sum of a pair of haplotypes is their corresponding genotype, i.e., $g_{i,k} = h_{j, k} + h_{l, k}$ for all $k\in\{1,\dots ,m\}$ iff genotype $g_i$ corresponds to pair of haplotypes $(h_{j}, h_{l})$.

The PolyIP encoding models the $2n$ haplotypes $h_i$, using $2nm$ binary variables $h_{i,j}$ where $1\leq i \leq 2n$ and $1\leq j \leq m$.
The haplotypes $h_{2i}$ and $h_{2i-1}$ are constrained by the encoding to resolve genotype $g_i$ by including for each $i\in\{1, \dots , n\}$ and each $j\in\{1,\dots ,m\}$ the constraint
$$ h_{2i,j} + h_{2i-1,j} = g_{i,j}. $$
The encoding contains binary variables $d_{i,j}$ where $1\leq i<j\leq 2n$ to detect whether $h_i$ and $h_j$ are different.
The constraints 
$$ \text{for all }k\in\{1,\dots ,m\}:\quad  d_{i,j} \geq h_{i,k}-h_{j,k} $$
and
$$ \text{for all }k\in\{1,\dots ,m\}:\quad  d_{i,j} \geq h_{j,k}-h_{i,k} $$
ensure that $d_{i,j}$ is set to 1 whenever $h_i\neq h_j$.

To compute the number of distinct haplotypes in the collection of $2n$ haplotypes variables $x_i$ with $1\leq i\leq 2n$ are included.
The constraints
$$ \text{for all }i\in\{1,\dots ,2n\}:\quad  x_{i} \geq 2 - i \sum_{j=1}^{i-1} d_{i,j} $$
ensure that $x_i$ is set to 1 whenever haplotype $h_i$ is distinct from haplotypes $h_1,\dots , h_{i-1}$.
The sum $\sum_{1\leq i\leq 2n} x_i$ then corresponds to the number of distinct haplotypes used.
The objective of the encoding is therefore to minimize $\sum_{1\leq i\leq 2n} x_i$.
The PolyIP encoding is presented in Figure~\ref{fig:enc-polyip}.

\begin{figure}
\begin{mdframed}
\centering
Minimize $\sum_{1\leq i \leq 2n} x_i$
\begin{align}
\text{for all  }i\in\{1,\dots , n\},\ j\in\{1,\dots ,m\}:&\quad h_{2i,j} + h_{2i-1,j} = g_{i,j} \\
\text{for all  }i, j\in\{1,\dots , 2n\},\text{ such that } i<j:&\quad d_{i,j} \geq h_{i,k}-h_{j,k} \\
&\quad d_{i,j} \geq h_{j,k}-h_{i,k} \\
\text{for all  }i\in\{1,\dots , 2n\}:&\quad x_{i} \geq 2 - i +  \sum_{j=1}^{i-1} d_{i,j} \\
\text{for all  }i,j\in\{1,\dots , 2n\},\ k\in\{1,\dots ,m\}:&\quad x_i\in\{0,1\} \\
&\quad h_{i,k}\in\{0,1\} \\
&\quad d_{i,j}\in\{0,1\}
\end{align}
\end{mdframed}
\caption{The PolyIP encoding.}
\label{fig:enc-polyip}
\end{figure}

Brown and Harrower additionally introduce optimizations to speed up solving the PolyIP formulation.
They note that actually an optimal solution of PolyIP may allow some $d_{i,j}$ variables to be set to 1 even though both $h_i$ and $h_j$ are the same.  
This results in the $y_{i,k}$ variables taking on fractional values in the LP relaxation of PolyIP degrading solving performance.
They suggest two solutions: 1) slightly perturbing the objective function to bias the solver towards setting $d_{i,j}$ to 0, 
and 2) adding constraints that enforce $d_{i,j}=0$ if $h_i=h_j$.
Other optimizations suggested include problem specific \emph{cuts} based on the input data as well as transitivity constraints for the $d_{i,j}$ variables.

\section{Discussion}

We have now presented three encodings for solving the HIPP problem, one SAT encoding (HIPP-SAT) and two ILP encodings (RTIP and PolyIP).
These encoding differ both in terms of their structure as well as their size, performance and available optimizations.
The HIPP-SAT and PolyIP encodings are structurally similar since they both utilize a structured representation of the haplotypes thus giving the solvers access to this representation.
In contrast the RTIP encoding uses an atomic representation of the haplotypes and computes necessary information about the relationship of the haplotypes and genotypes in a pre-processing phase.

This structural difference is reflected in the sizes of the encodings.
Let $n$ be the number of genotypes, $m$ the number of loci and $r$ the number of distinct haplotypes used in HIPP-SAT.
The size of HIPP-SAT scales polynomially in $n$, $m$ and $r$, and the size of PolyIP is bounded by a polynomial in $n$ and $m$.
The size of RTIP, however, scales linearly with respect to $n$, but exponentially with respect to the number of heterozygous loci (which is bounded above by $m$).
That is, HIPP-SAT and PolyIP have polynomial size whereas the size of RTIP scales exponentially.

The RTIP encoding is relatively straightforward since part of the work is done in the pre-processing phase.
Due to this fact the optimizations used for RTIP mostly concern the pre-processing phase.
Gusfield~\cite{DBLP:conf/cpm/Gusfield03} explains how the number of haplotypes to unfold can be reduced and how the unfolding can be accomplished efficiently.
The optimizations utilized in the HIPP-SAT and PolyIP formulations, on the other hand, modify the plain encoding itself by, e.g., including symmetry-breaking or transitivity constraints.

Brown and Harrower note that their PolyIP encoding is not as fast as Gusfield's RTIP over instances that RTIP can solve~\cite{DBLP:journals/tcbb/BrownH06}.
Their PolyIP formulation, however, scales to larger instances due to its polynomial size (as opposed to the exponential size of RTIP).
Due to these observations Brown and Harrower also formulate a hybrid encoding \emph{HybridIP} that is a mix between PolyIP and RTIP, and which can solve instances fast and scales to larger instances than RTIP~\cite{DBLP:conf/cpm/Gusfield03}.
Essentially the HybridIP uses a mixed representation for the haplotypes; 
some haplotypes are unfolded and have atomic representation as in RTIP whereas others have a structured representation like that of PolyIP.
However, the performance and scalability of SAT-based methods for HIPP (such as HIPP-SAT presented here) exceeds those of both PolyIP and RTIP~\cite{DBLP:journals/jcb/GracaLMO10}.


In addition to the SAT and ILP encodings for the HIPP problem presented here, there are still other suggested algorithms and encodings proposed for the problem.
In~\cite{DBLP:journals/jcb/GracaLMO10} Graca et al provide a survey of proposed methods for solving the HIPP problem.
Wang and Xu, on the other hand, formulated a branch-and-bound algorithm~\cite{DBLP:journals/bioinformatics/WangX03a}.
Graca et al~\cite{DBLP:journals/anor/GracaMLO11} propose and formulate methods utilizing \emph{pseudo-boolean optimization} 
whereas Esra et al~\cite{DBLP:conf/lpnmr/ErdemET09} propose the use of \emph{answer-set programming} for solving the HIPP problem.
In~\cite{DBLP:journals/jda/JagerCZ16} JÃ¤ger et al study a generalization of HIPP, namely the problem of generating all optimal solutions to the HIPP problem.




\bibliographystyle{plain}
\bibliography{paper.bib}

\end{document}
