\documentclass[12pt,a4paper]{article}
%\usepackage[margin=1in]{geometry}

% Localization and encoding
\usepackage[utf8]{inputenc}
\usepackage[english]{babel}
\usepackage{csquotes}

% Math
\usepackage{amsthm}
\usepackage{amsmath}
\usepackage{amsfonts}

% Graphics
%\usepackage{rotating}
\usepackage{graphicx}
\graphicspath{ {./resources/images/} }
\usepackage{multirow}

% algorithms
%\usepackage{algorithm}
%\usepackage{algorithmic}
\usepackage[algoruled]{algorithm2e}

\usepackage{pdfpages}
\usepackage{multirow}
\usepackage{placeins}
\usepackage{hyperref}
\usepackage{caption}
\usepackage{tikz}
\usetikzlibrary{math}
\usetikzlibrary{calc}
\usetikzlibrary{arrows.meta}



% Counters
\newcounter{thmCounter}
\newcounter{defCounter}

% Theorem environments
\newtheorem{observation}{\bf Observation}
\newtheorem{theorem}{\bf Theorem}
\newtheorem{lemma}{\bf Lemma}
\newtheorem{prop}{\bf Proposition}
\newtheorem{rmk}{\bf Remark}
\newtheorem{definition}{\bf Definition}
\newtheorem{example}{\bf Example}
\newtheorem{corollary}{\bf Corollary}

\newcommand{\TODO}[1]{\textcolor{red}{#1}}



\sloppy
\title{Comparing SAT and ILP formulations for Computational and Systems Biology problems}
\date{2020 Autumn}
\author{Jarkko Savela}

\begin{document}
\pagenumbering{gobble}
\maketitle
\newpage
\tableofcontents
\newpage
\pagenumbering{arabic}

\section{Introduction}


\section{Boolean Satisfiability}
Boolean satisfiability~\cite{DBLP:series/faia/2009-185}, often referred to as the SAT problem, is the satisfiability problem of propositional logic.
That is, the problem is to find an assignment $s$ that satisfies the input formula $\phi$.
The formula $\phi$ is inductively constructed from propositional atoms (i.e., Boolean variables) and the usual connectives; conjunction $\wedge$, disjunction $\vee$, negation $\neg$, implication $\rightarrow$ and bi-implication $\leftrightarrow$.
The assignment $s$ gives each boolean variable in $\phi$ a truth value, 0 or 1.

In practice the input formula is usually restricted to be in \emph{conjunctive normal form} (CNF).
A formula $\phi$ is in CNF if it is a conjunction of \emph{clauses}
$$ \bigwedge_{i\in I} C_i $$
where each clause $C_i$ is a disjunction of \emph{literals}
$$ C_i = \bigvee_{j\in J_i} l_j $$
and each literal is either an atom $p$ or a negated atom $\neg p$.
CNF is a convenient standard format partly due to the fact that efficient polynomial-time algorithms exist for transforming arbitrary formulas into \emph{linear-size, equisatisfiable} CNF formulas~\cite{tseitin1983, DBLP:journals/jsc/PlaistedG86}.

Practically efficient algorithms for solving the SAT problem have been developed despite the fact that SAT is an NP-Complete problem~\cite{DBLP:conf/stoc/Cook71}, 
and SAT has become an importent problem solving technique with applications in model checking and verification, planning, scheduling, cryptography, computational biology et cetera.
The most efficient current SAT solvers are based on the \emph{conflict-driven clause learning} algorithm (CDCL) and use many optimizations and efficient data structures~\cite{DBLP:conf/iccad/SilvaS96, DBLP:journals/tc/Marques-SilvaS99, DBLP:conf/dac/MoskewiczMZZM01, DBLP:conf/aaai/GomesSK98, DBLP:journals/dam/GoldbergN07, ryan2004efficient, DBLP:conf/aaai/BayardoS97, DBLP:conf/sat/LewisSB05}.


\section{Integer Linear Programming}
Integer linear programming (ILP) is the problem of finding an optimal assignment to a set of integer variables subject to constraints in the form of linear inequalities~\cite{DBLP:books/ph/PapadimitriouS82}.
An ILP program thus consists of a vector of integer variables $\mathbf{x}$, a cost vector $\mathbf{c}$ as well as a matrix $A$ and vector $\mathbf{b}$ of coefficients with the goal of minimizing $\mathbf{c}^T \mathbf{x}$ under the constraints that $A\mathbf{x} \leq \mathbf{b}$.

ILP is also an NP-hard problem~\cite{DBLP:conf/coco/Karp72} and finds applications in planning, scheduling, computational biology and in many other domains.

The \emph{simplex algorithm} is a practically efficient (although worst-case exponential time) algorithm for solving general linear programs where the variables may take non-integral values as well.
ILPs can be solved via a \emph{branch-and-bound} procedure in which the simplex algorithm is used to solve the continuous relaxations of the program with or without the use of \emph{cutting planes}.
Branch-and-bound procedure utilizing cutting planes, i.e., \emph{branch-and-cut}, uses auxiliary constraints to prune away non-integral solutions~\cite{DBLP:books/ph/PapadimitriouS82}.


\section{Protein folding}


\subsection{SAT formulation}

\subsection{ILP formulation}


\section{Haplotype inference}
The human genetic information is contained in the double-stranded DNA molecules consisting of a linear sequence of 4 bases (\textbf{A}denosine, \textbf{C}ytosine, \textbf{G}uanine, \textbf{T}hymine) in the phosphate-deoxyribose backbone.
The dna molecules, which are packed into chromosomes, consist of protein-coding intervals, which are called genes, and other non-coding parts.
Different variants of a gene are called alleles.
Humans are a \emph{diploid} species, which means that we have two copies of each chromosome (one from each parent).
We thus have two alleles of each gene, and we are said to be \emph{homozygous} with regard to a gene if we have two identical alleles and \emph{heterozygous} otherwise.

\emph{Haplotype} refers to the information contained in a single chromosome whereas the conflated information contained in the pair of  homologous chromosomes inherited is referred to as the \emph{genotype}.
In practice it is difficult and expensive to experimentally detect haplotype information.
Detecting genotype information has, however, become much faster, cheaper and more reliable in the recent decades.
Potential haplotypes can be computationally inferred from genotype data, and this is the problem referred to as the haplotype inference (HI) problem.
Haplotype data has the potential to yield new insight into genetic predisposition to, e.g., diseases, which makes the HI problem of great importance.

\subsection{Haplotype Inference by Pure Parsimony}
The genotype data for the computational problem is represented as vectors over the alphabet $\{0, 1, 2\}$. 
The values $0$ and $1$ indicate that the location in the diploid genotype is homozygous with allele $0$ or allele $1$, respectively, 
whereas the value $2$ indicates that the location is heterozygous.
The corresponding haplotype data then consists of a pair of vectors over the binary alphabet $\{0, 1\}$ with each pair corresponding to the haplotype inherited from each parent.
Denote a genotype vector by $g=g_1\dots g_n$ and the haplotype vectors $h=h_1\dots h_n$ and $h'=h'_1\dots h'_n$.
The pair of haplotype vectors $(h, h')$ corresponds to genotype $g$ if 
\begin{itemize}
\item $h_i=h'_i=0$ whenever $g_i = 0$,
\item $h_i=h'_i=1$ whenever $g_i = 1$, and
\item $h_i \neq h'_i$ whenever $g_i = 2$.
\end{itemize}
The haplotype $(01011, 00001)$, for example, corresponds to the genotype $02021$.
Note that there can be several distinct pairs of haplotype vectors for each genotype.
For example the genotype $02021$ corresponds also to the pair $(00011, 01001)$.

The input to the HI problem is the genotype data of $m$ individuals with the genotype vectors being all of length $n$.
The objective is to find a pair of haplotype vectors for each of the $m$ genotypes such that the two haplotype vectors yield exactly the genotype vector.
That is, for each genotype vector $g$ find a pair of haplotype vectors $(h, h')$ such that $g$ corresponds to $(h, h')$.
The Haplotype Inference by Pure Parsimony (HIPP) sets another constraint on this problem, namely that the set of distinct haplotype vectors used should be as small as possible.
This condition has been experimentally validated to yield relatively accurate haplotype data~\cite{DBLP:journals/bioinformatics/WangX03a}.

The HIPP problem was shown by Lancia et al to be APX-hard~\cite{DBLP:journals/informs/LanciaPR04}.

In the following we present two solutions for the HIPP problem: a SAT encoding by Lynce and Marques-Silva~\cite{DBLP:conf/aaai/LynceM06} and an ILP formulation by Gusfield~\cite{DBLP:conf/cpm/Gusfield03}.

\subsection{SAT formulation}

Denote by $\mathcal{G}=\{g_i\mid 1\leq i \leq n\}$ a set of $n$ genotype vectors of length $m$ and by $g_{ij}$ the $j$'th entry of genotype vector $g_i$.
We construct a propositional formula $\phi (\mathcal{G}, r)$ which is satisfiable if and only if there exists a set of haplotypes $\mathcal{H}$ of size $r$ such that each genotype $g_i\in \mathcal{G}$ is explained by some pair fo haplotypes $(h_j, h_k)$ in $\mathcal{H}$.
To solve the HIPP problem the satisfiability of $\phi (\mathcal{G}, r)$ is checked for increasing values of $r$ until a value is found such that $\phi (\mathcal{G}, r)$ is satisfiable but $\phi (\mathcal{G}, r-1)$ is not.
This value of $r$ is then the minimum number of haplotypes required, and the actual haplotypes can be extracted from the satisfying assignment to $\phi (\mathcal{G}, r)$.

The formula $\phi (\mathcal{G}, r)$ considers a set of $r$ distinct haplotypes $h_k$ and encodes the assignment of a pair $(h_{k_1}, h_{k_2})$ to each $g_i$ using Boolean variables $s^a_{k_1i}$, $s^b_{k_2i}$ and $h_{kj}$.
That is, variables $s^a_{k_1i}$ and $s^b_{k_2i}$ being set true indicates that $(h_{k_1}, h_{k_2})$ is the pair of haplotypes assigned to genotype $g_i$.
The variable $h_{kj}$ is the $j$'th entry of haplotype $h_k$.
The formula then requires two types of constraints
\begin{enumerate}
\item[(i)] The selector variable $s^a_{ki}$ should select a unique haplotype, i.e., for each genotype $g_i$ exactly one variable $s^a_{ki}$ should be set true (and similarly for $s^b_{ki}$). 
\item[(ii)] Each genotype $g_i$ should be explained by its selected pair of haplotypes $(h_{k_1}, h_{k_2})$. 
\end{enumerate}

Constraint (i) corresponds to equations $\sum_{1\leq k\leq r} s^a_{ki} = 1$ and $\sum_{1\leq k\leq r} s^b_{ki} = 1$ holding for all $i$ such that $1\leq i \leq n$.
This constraint is enforced in $\phi (\mathcal{R}, r)$ utilizing a propositional encoding of an \emph{1-bit adder circuit}.
The adder circuit uses variables $v^a_{ki}$ where $1\leq k \leq r+1$ and $1\leq i\leq n$ and the following constraints (the ones for $s^b_{ki}$ are similar).
\begin{eqnarray}
\forall i\text{ such that } 1\leq i\leq n: & v^a_{1i} \leftrightarrow s^a_{1i}\label{hipp-sat:clause1}\\
\forall k, i\text{ such that } 1\leq k \leq r,\ 1\leq i\leq n: & \neg v^a_{ki} \vee \neg s^a_{ki}\label{hipp-sat:clause2}\\
\forall k, i\text{ such that } 1\leq k \leq r,\ 1\leq i\leq n: & v^a_{(k+1)i} \leftrightarrow (s^a_{ki}\vee v^a_{ki} )\label{hipp-sat:clause3}\\
\forall i\text{ such that } 1\leq i\leq n: & v^a_{(r+1)i}=1\label{hipp-sat:clause4}
\end{eqnarray}
Constraint \eqref{hipp-sat:clause1} enforces $v^a_{1i}$ and $s^a_{1i}$ to take on the same value in every satisfying assignment whereas constraint \eqref{hipp-sat:clause4} enforces variable $v^a_{(r+1)i}$ to be evaluated to true.
Constraint \eqref{hipp-sat:clause2} prevents $v^a_{ki}$ and $s^a_{ki}$ from being simultaneously assigned true, 
and constraint \eqref{hipp-sat:clause3} sets $v^a_{(k+1)i}$ to 1 whenever either $s^a_{ki}$ or $v^a_{ki}$ is assigned 1.

It can be shown using induction that the variable $v^a_{ki}$ (for $k>1$) represents the disjunction $\bigvee_{1\leq h\leq k-1}s^a_{hi} = s^a_{1i}\vee\dots \vee s^a_{(k-1)i}$.
From constraints \eqref{hipp-sat:clause1} and \eqref{hipp-sat:clause3} it follows that $v^a_{2i}$ is equivalent to $s^a_{1i}$.
Constraint \eqref{hipp-sat:clause3} also provides yields the proof of the inductive step.
Thus, constraint \eqref{hipp-sat:clause4} which sets $v^a_{(r+1)i}=1$ ensures that at least one of $s^a_{1i},\dots , s^a_{ri}$ is set to 1.
Furthermore, there can only be one $s^a_{ki}$ set to 1 due to the following reasoning.
Since $v^a_{ki}$ is equivalent to $\bigvee_{1\leq h\leq k-1}s^a_{hi} = s^a_{1i}\vee\dots \vee s^a_{(k-1)i}$ it follows that $v^a_{ki}$ being set 1 enforces all $v^a_{hi}$ to be set 1 for $h>k$.
Now assume there are two indices $h$ and $h'$ ($h<h'$) such that both $s^a_{hi}$ and $s^a_{h'i}$ are assigned true.
By the previous observation $v^a_{h'i}$ is also assigned true which violates constraint \eqref{hipp-sat:clause2}.

Next we formalize constraint (ii) stating that each genotype $g_i$ should be explained by its selected pair of haplotypes $(h_{k_1}, h_{k_2})$.
If $g_{ij}$ (the $j$'th entry of $g_i$) is 0, then the $j$'th entries of the chosen haplotypes should also be 0.
This is encoded by $s^a_{ki}\rightarrow\neg h_{kj}$ and $s^b_{ki}\rightarrow\neg h_{kj}$ for all $1\leq k \leq r$.
Similarly, if $g_{ij}$ is 1, then the $j$'th entries of the chosen haplotypes should also be 1, which is encoded by $s^a_{ki}\rightarrow h_{kj}$ and $s^b_{ki}\rightarrow h_{kj}$ for all $1\leq k \leq r$.
If $g_{ij}$ is 2, then the $j$'th entry of the haplotypes should differ.
Variables $g^a_{ij}$ and $g^b_{ij}$ are used to decide which haplotype should be assigned 0 and which should be assigned 1.
The clauses $(g^a_{ij}\vee g^b_{ij})\wedge(\neg g^a_{ij}\vee \neg g^b_{ij})$ enforce $g^a_{ij}$ and $g^b_{ij}$ to take on different values.
The constraints $s^a_{ki}\rightarrow (h_{kj}\leftrightarrow g^a_{ij})$ and $s^b_{ki}\rightarrow (h_{kj}\leftrightarrow g^b_{ij})$ then enforce the selected haplotypes to take on values dictated by $g^a_{ij}$ and $g^b_{ij}$.
\subsection{ILP formulation}

\bibliographystyle{plain}
\bibliography{paper.bib}

\end{document}
